### ISSUE TITLE
Use of methods on dynamically loaded instances of abstract base classes raise `reportAbstractUsage` errors

### ISSUE THREAD

**Issue by joel-wright** [microsoft/pyright#7105]
In our project we dynamically load implementations of abstract base classes at runtime based on config. Since pyright `1.1.345` all uses of methods labelled as abstract on the base class of those dynamically loaded instances are detected as calls to unimplemented abstract methods and flagged as errors.

I have tried to provide a minimal reproduction here:

```python
from abc import ABC, abstractmethod
from typing import Type, TypeVar


class ABase(ABC):
    """Base class"""

    @abstractmethod
    def some_a(self) -> int:
        """Simple method"""
        raise NotImplementedError


class AImpl(ABase):
    """Implementation of ABase"""

    def __init__(self):
        self.a = 10

    def some_a(self) -> int:
        """Implementation of simple method"""
        return self.a


T = TypeVar("T")
LOOKUP_FACTORY = {"ABase": AImpl}


def get_me_an_object(clz: Type[T]) -> T:
    if clz.__name__ in LOOKUP_FACTORY:
        factory = LOOKUP_FACTORY[clz.__name__]
        ret = factory()
        if not isinstance(ret, clz):
            raise RuntimeError("wrong type from factory")
        return ret
    raise RuntimeError("Unknown")


def run():
    d_a = get_me_an_object(ABase)
    print(d_a.some_a())


if __name__ == "__main__":
    run()
```

Pyright versions `<=1.1.344` are happy with the above example, but versions `>=1.1.345` raise an error on the use of `some_a()`:

```
testcase.py:41:11 - error: Method "some_a" cannot be called because it is abstract and unimplemented (reportAbstractUsage)
```

**Comment by erictraut** [comment]
Thanks for the bug report. This was a long-standing bug that was previously masked but was exposed by a recent change. It specifically affects the case where a reference to an abstract class (like `ABase` in your example) is passed through a generic function that accepts a `type[T]` and returns a `T` (like in your example). The set of conditions was quite specific, which also explains why I didn't have coverage for this in our test suite. 

I've fixed the underlying bug and added a test case to prevent future regressions like this. The fix will be in the next release of pyright.

**Comment by joel-wright** [comment]
Thanks for the update & the quick fix!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.349, which I just published.
